File: backend/app/schemas/market.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel

from app.domain.enums import InstrumentType


class InstrumentOut(BaseModel):
    id: uuid.UUID
    symbol: str
    name: str
    type: InstrumentType
    exchange: str | None = None
    sector: str | None = None
    currency: str = "USD"


class SearchResponse(BaseModel):
    results: list[InstrumentOut]


class QuoteOut(BaseModel):
    symbol: str
    price: float
    change: float | None = None
    change_pct: float | None = None
    open: float | None = None
    high: float | None = None
    low: float | None = None
    prev_close: float | None = None
    volume: float | None = None
    as_of: datetime | None = None


class CandleOut(BaseModel):
    t: datetime
    o: float
    h: float
    l: float  # noqa: E741
    c: float
    v: float | None = None


class CandleSeriesOut(BaseModel):
    symbol: str
    resolution: str
    candles: list[CandleOut]

File: web/src/models/market.ts
import type { InstrumentType } from "./enums";

export interface Instrument {
  id: string;
  symbol: string;
  name: string;
  type: InstrumentType;
  exchange?: string | null;
  sector?: string | null;
  currency: string;
}

export interface Quote {
  symbol: string;
  price: number;
  change?: number | null;
  change_pct?: number | null;
  open?: number | null;
  high?: number | null;
  low?: number | null;
  prev_close?: number | null;
  volume?: number | null;
  as_of?: string | null;
}

export interface Candle {
  t: string;
  o: number;
  h: number;
  l: number;
  c: number;
  v?: number | null;
}

export interface CandleSeries {
  symbol: string;
  resolution: string;
  candles: Candle[];
}

export interface SearchResponse {
  results: Instrument[];
}
